File: web-frontend/src/data/repositories/AdRepositoryImpl.ts
import { Ad } from '../../domain/entities/Ad';
import { IAdRepository } from '../../domain/repositories/IAdRepository';
import { AdDataSource } from '../datasources/AdDataSource';

export class AdRepositoryImpl implements IAdRepository {
  constructor(private dataSource: AdDataSource) {}

  async getAds(): Promise<Ad[]> {
    return await this.dataSource.getAds();
  }

  async getAdById(id: string): Promise<Ad | null> {
    return await this.dataSource.getAdById(id);
  }

  async getAdsByType(type: string): Promise<Ad[]> {
    const ads = await this.dataSource.getAds();
    return ads.filter(ad => ad.type === type);
  }

  async createAd(ad: Omit<Ad, 'id' | 'createdAt' | 'updatedAt'>): Promise<Ad> {
    return await this.dataSource.createAd(ad);
  }

  async updateAd(id: string, ad: Partial<Ad>): Promise<Ad> {
    return await this.dataSource.updateAd(id, ad);
  }

  async deleteAd(id: string): Promise<void> {
    await this.dataSource.deleteAd(id);
  }
}

File: android-sdk/samplesdk/src/main/java/com/example/android_sdk/data/repositories/AdRepositoryImpl.kt
package com.example.android_sdk.data.repositories

import com.example.android_sdk.domain.entities.Ad
import com.example.android_sdk.domain.repositories.AdRepository
import com.example.android_sdk.data.datasources.AdDataSource

class AdRepositoryImpl(private val dataSource: AdDataSource) : AdRepository {
    override suspend fun getAds(): List<Ad> {
        return dataSource.getAds()
    }

    override suspend fun getAdById(id: String): Ad? {
        return dataSource.getAdById(id)
    }

    override suspend fun getAdsByType(type: String): List<Ad> {
        val ads = dataSource.getAds()
        return ads.filter { it.type.name.lowercase() == type.lowercase() }
    }

    override suspend fun createAd(ad: Ad): Ad {
        return dataSource.createAd(ad)
    }

    override suspend fun updateAd(id: String, ad: Ad): Ad {
        return dataSource.updateAd(id, ad)
    }

    override suspend fun deleteAd(id: String) {
        dataSource.deleteAd(id)
    }
}
